crates/noirc_evaluator/src/ssa/optim.rs

```diff
@@ -1,225 +1,151 @@
         }
     }
+
+    None
 }
 
-fn evaluate_intrinsic(
-    irgen: &mut SsaContext,
-    op: acvm::acir::OPCODE,
-    lhs: FieldElement,
-    rhs: FieldElement,
-) -> NodeId {
+fn evaluate_intrinsic(ctx: &mut SsaContext, op: acvm::acir::OPCODE, args: Vec<u128>) -> NodeEval {
     match op {
         acvm::acir::OPCODE::ToBits => {
-            let lhs_int = lhs.to_u128();
-            let rhs_int = rhs.to_u128() as u32;
-            let a =
-                irgen.mem.create_new_array(rhs_int, node::ObjectType::Unsigned(1), &String::new());
-            let pointer = node::Variable {
+            let bit_count = args[1] as u32;
+            let array_id = ctx.mem.create_new_array(bit_count, ObjectType::Unsigned(1), "");
+            let pointer = Variable {
                 id: NodeId::dummy(),
-                obj_type: node::ObjectType::Pointer(a),
+                obj_type: ObjectType::Pointer(array_id),
                 root: None,
                 name: String::new(),
                 def: None,
                 witness: None,
-                parent_block: irgen.current_block,
+                parent_block: ctx.current_block,
             };
-            for i in 0..rhs_int {
-                if lhs_int & (1 << i) != 0 {
-                    irgen.mem.arrays[a as usize]
-                        .values
-                        .push(InternalVar::from(FieldElement::one()));
+
+            for i in 0..bit_count {
+                if args[0] & (1 << i) != 0 {
+                    ctx.mem[array_id].values.push(InternalVar::from(FieldElement::one()));
                 } else {
-                    irgen.mem.arrays[a as usize]
-                        .values
-                        .push(InternalVar::from(FieldElement::zero()));
+                    ctx.mem[array_id].values.push(InternalVar::from(FieldElement::zero()));
                 }
             }
-            irgen.add_variable(pointer, None)
+
+            let new_var = ctx.add_variable(pointer, None);
+            NodeEval::VarOrInstruction(new_var)
         }
         _ => todo!(),
     }
 }
-
 ////////////////////CSE////////////////////////////////////////
 
 pub fn find_similar_instruction(
     igen: &SsaContext,
-    lhs: NodeId,
-    rhs: NodeId,
-    prev_ins: &VecDeque<NodeId>,
-) -> Option<NodeId> {
-    for iter in prev_ins {
-        if let Some(ins) = igen.try_get_instruction(*iter) {
-            if ins.lhs == lhs && ins.rhs == rhs {
-                return Some(*iter);
-            }
-        }
-    }
-    None
-}
-
-pub fn find_similar_instruction_with_multiple_arguments(
-    igen: &SsaContext,
-    lhs: NodeId,
-    rhs: NodeId,
-    ins_args: &[NodeId],
+    operation: &Operation,
     prev_ins: &VecDeque<NodeId>,
 ) -> Option<NodeId> {
     for iter in prev_ins {
         if let Some(ins) = igen.try_get_instruction(*iter) {
-            if ins.lhs == lhs && ins.rhs == rhs && ins.ins_arguments == ins_args {
+            if &ins.operator == operation {
```

Comment:
This potentially changes the behavior as the operation contains more than the arguments.

Reply:
I think this part needs to be refactored but I am OK to do this later on.